File: Examples_VI_CtCI_C/Example 15/Example15.c
#include <stdio.h>
//not finished

int fib (int n, int memo[]) {
	if (n <= 0) return 0;
	else if (n == 1) return 1;
	else if (memo[n] > 0) return memo[n];
	
	memo[n] = fib(n - 1, memo) + fib(n - 2, memo);
	return memo[n];
}

void allFib (int n) {
	int memo[n + 1];
	int i;
	
	for (i = 0; i <= n; i++) {
		printf("%d: %d\n", i, fib(i, memo));
	}
}

int main() {
	int n;
	allFib(10);
}

File: Examples_VI_CtCI_Python/Example15.py
def allFib(n):
    memo = [0] * (n + 1)
    i = 0
    while (i <= n):
        print(str(i) + ": " + str(fib(i, memo)))
        i += 1

def fib(n, memo):
    if (n <= 0):
        return 0
    elif (n == 1):
        return 1
    elif (memo[n] > 0):
        return memo[n]
    memo[n] = fib(n - 1, memo) + fib(n - 2, memo)
    return memo[n]

n = 20
allFib(n)